011-kafka-canary.md

```diff
@@ -0,0 +1,69 @@
+# Strimzi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and periodically producing and consuming events on the topics. 
+
+## Current situation
+
+Currently Strimzi does not have any canary style feature. It does have health checks for Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following in indication that the Kafka cluster is operating as expecting from a customers perspective. I.e. messages can successfully be produced and consumed. Some would be exported to reflect the canary's activity. Initial metrics would included:
+
+  * records-produced-total
+  * records-consumed-total
+  * produce-error
+  * consume-error
+  * latency
+
+In future the canary can be used by the Kafka Roller to improve health checks. However this is beyond of scope of the current proposed work.
+
+## Proposal
+
+The canary would be run as separate pod alongside Strimzi operator. It would be specified (optionally) in the Kafka custom resource and deployed by the Strimzi controller. The code will live in a separate repository within the Strimzi project. This is similar setup to the Kafka Bridge. 
+
+The canary will be built using Golang.
+
+There are two main options for how the topic is configured:
+
+### Option 1: Partition over replicas
+Create a topic with 1 partition per broker node and a replication factor of zero. The canary would produce/consume from each partition. 
+
+#### Pros
+
+  * Errors will indicate that a specific broker is unavailable.
+
+#### Cons
+
+  * Brokers are expected to be unavailable during rolling updates after cluster configuration changes. 
```

Comment:
Well this is also a pro, since being able to produce and consume again tells you that the broker came back OK (e.g. there weren't problems mounting volumes etc). It's only a con because on its own the canary can't tell that a broker was intentionally being rolled. Presumably this could be dealt with in prometheus alerting by exporting a metric from the CO for each broker which was 1 when intentionally restarting and 0 otherwise, allowing the difference with the canary availability metric to indicate a problem.